Feature/jogger

skaterzero807 opened the pull request:
## Summary
Adding jogging functionality. Allows a user to attach an analog joystick to the smoothieboard and control the speed of different axes on their machine with it. Useful for setting up a job on a CNC, 3D printer or laser cutter.

The code consists of two separate modules:
- Joystick: handles the reading of a joystick (analog device), and mapping the raw ADC to useful values
- Jogger: reads an object such as a joystick and controls the speed and direction the machine moves

## Documentation
Documentation which will accompany these modules:
- Joystick setup/configs: http://smoothieware.org/joystick
- Jogger setup/configs: http://smoothieware.org/jogger
- Joystick developer doc: http://smoothieware.org/joystick-dev
- Jogger developer doc: http://smoothieware.org/jogger-dev

## Testing
- **Tested** on steppers connected to smoothie in cartesian mode, not yet tested on an actual cartesian CNC
- Not yet tested on Delta, H-bot, SCARA, etc.
- Not yet tested with endstops
- Not yet tested with 3D printers, maybe could do extruder axis control with jogging?

skaterzero807 committed:
New Joystick feature

Adding old files from the test build of a Joystick feature. Code added
in this commit was not previously on Github and is being added now. It
is possible that changes to the edge branch since this code was written
make it no longer functional.

skaterzero807 committed:
Ignoring visual studio project folder in gitignore

Added line to gitignore to ignore changes to the Visual Studio project
containing the smoothie source.

skaterzero807 committed:
Test, commit deletions for -rewrite branch

Testing to see if commiting deletions to the -rewrite branch leaves the
original branch unharmed

skaterzero807 committed:
Basic JoystickAxis functionality added

Added functionality for different joystick axis modules (via
JoystickAxisPool).
Added functionality for other modules to request the position of an axis
through a Public Data Request (via JoystickPublicAccess)
Removed unneeded references to headers, cleaned out unused code.

skaterzero807 committed:
Adding small debug outputs, removed ADC filtering

ADC filtering removed from JoystickAxis.cpp when reading the ADC value.
ADC now performs the filtering somehow.

skaterzero807 committed:
More debug features, cleaning up

Cleaned up more references to ADC filtering
Added debugging features for a public data request
Added raw ADC reading to public access data structure

skaterzero807 committed:
Minor bug fixes

Fixed slow timer attach bug (frequency not interval)
Fixed unsigned int bug when subtracting
Fixed ADC auto-reading to reject values which are different from the
previous

skaterzero807 committed:
Introduction of Jogger files

Skeleton files for the Jogger module

skaterzero807 committed:
Introduction of Jogger files

Skeleton files for the Jogger module

skaterzero807 committed:
Clean up public data request debug code

Public data request debugging code removed, now it is stored in the
Jogger module and will be tested from there.

skaterzero807 committed:
Clean up public data request debug code

Public data request debugging code removed, now it is stored in the
Jogger module and will be tested from there.

skaterzero807 committed:
Basic Jogger functionality

Jogger files now contain basic code to load settings and read from a
joystick.

skaterzero807 committed:
Basic Jogger functionality

Jogger files now contain basic code to load settings and read from a
joystick.

skaterzero807 committed:
Merge pull request #1 from skaterzero807/feature/Joystick-rewrite

Joystick feature rewrite

skaterzero807 committed:
Merge branch 'feature/Jogger' of https://github.com/skaterzero807/Smoothieware into feature/Jogger

skaterzero807 committed:
Added comments to header parameters

skaterzero807 committed:
Changed ADC readings into volts, updated refresh_interval to _rate

Changes made to return raw ADC values in voltage so the configuration is
more intuitive
Also updated refresh_interval to be refresh_rate to make config more
intuitive

skaterzero807 committed:
Changed ADC readings into volts, updated refresh_interval to _rate

Changes made to return raw ADC values in voltage so the configuration is
more intuitive
Also updated refresh_interval to be refresh_rate to make config more
intuitive

skaterzero807 committed:
Change refresh_interval to refresh_rate

Change the jogger speed update parameter to make the configuration more
intuitive

skaterzero807 committed:
Testing Gcode-based jog, still freezes smoothie

Committing non-working code for Gcode-based jogging, does not work. Will
be deleting and trying a different method

skaterzero807 committed:
Testing delta-move based jogging

Commiting non-working delta-move jog, does not work as intended. Will
switch to manual control of steppers.

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/Smoothieware/edge' into feature/Jogger

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/Smoothieware/edge' into feature/Jogger

skaterzero807 committed:
add debug notes to jogger.cpp

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/Smoothieware/edge' into feature/Joystick

skaterzero807 committed:
revert to GCode based jogging

reverting to GCode based jogging as seen in
e78b1ea44c4374cf469813bd3430e56b06466ea4 recommended by @arthurwolf

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/Smoothieware/edge' into feature/Jogger

skaterzero807 committed:
additions to gitignore

skaterzero807 committed:
remove console printing for debugging

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/Joystick' into feature/Jogger

# Conflicts:
#	.gitignore
#	src/modules/tools/joystick/JoystickAxis.cpp

skaterzero807 committed:
remove accidental merge conflict text

skaterzero807 committed:
update jogger to maximum working sample

using small Gcode steps to acheive jogging
current code does not jog, but is most amount of code that will run
without a crash
L174 if uncommented causes a crash, will investigate

skaterzero807 committed:
update Jogger to minimum non-working sample

this is only a minor change from previous, but causes a crash when
jogging

skaterzero807 committed:
fix memory issue in Jogger

most likely cause of crashes in prev examples was memory leak, causing
the Gcode's command string to overflow the stack
possibly fixed using string on buffer, breaking up Gcodes for G90/91

skaterzero807 committed:
check M-code before printing debug into

change debug info to only print on M777 command, so when a G-code is
submitted the printing doesn't overwhelm the smoothie

skaterzero807 committed:
move Gcode actions to main loop

move stepper action out of a SlowTicker function to the main loop to
prevent crashes

skaterzero807 committed:
refactor to split gcode sends and joystick updates

move code for joystick analog read, speed mapping, state updating to a
SlowTicker, rather than run in main loop
add two states, one for joystick being moved, and another for robot
being jogged

skaterzero807 committed:
refactor for efficiency, readability

move abs/rel gcode into one-time command when enabling jogging
move all jogging state updates into main loop, slow tick update only
works on reading the joystick and updating the joystick's state

skaterzero807 committed:
update speed calc. for arbitrary num. of axes

changed speed calculation to use a loop so it works with any number of
defined axes
changed squaring from powf to simple multiplication (maybe faster?)

skaterzero807 committed:
update to use push/pop_state to restore settings

changed manual abs/rel state saving to use push/pop_state
removed debug print lines for start/stop jogging

skaterzero807 committed:
change step_scale_factor to step_frequency

change the mm per segment to be specified as the number of segments per
second. The mm per segment is now calculated by speed/freq. This means
the distance per segment changes as the jogger moves faster, taking
longer steps to ensure it doesn't run out of things to do before the
next command.

skaterzero807 committed:
bugfix in debug string

fix debug string for new segment frequency variable

skaterzero807 committed:
bugfix for unit error in segment length calc

skaterzero807 committed:
add M-code numbers to config

add ability to specify which M-codes will be used for jog axis set or
toggle

skaterzero807 committed:
update comments

skaterzero807 committed:
update to add variable machine axis mapping

add variable machine axis mappings to config
update Gcode generation to use custom machine axis mapping

skaterzero807 committed:
add code to handle setting machine jog axes

skaterzero807 committed:
cleanup of code

removed debug lines and extra includes
added "this->" to clarify class members

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/Joystick' into feature/Jogger

skaterzero807 committed:
clean-up for pull request

skaterzero807 committed:
clean-up for pull request

clean up for pull request

fix whitespace and gitignore for pull request

fix newline missing in gitignore

add joystick example config

fix joystick example config

skaterzero807 committed:
clean up for pull request

arthurwolf commented:
Thanks for the code ! 
I think we want to have users with diverse machines test this. Don't hesitate to post about it in the various Smoothie communities online.

skaterzero807 commented:
No problem, I would like to finish the documentation on the new wiki but
when it is done I'll put up announcements for beta testers.

wolfmanjm commented on .gitignore:
```diff
@@ -21,4 +21,4 @@ debug
 /win_install.log
 /linux_install.log
-/src/default_excludes.mk
+/src/default_excludes.mk
```
This should not be here please remove it as it did not change other than removing the last \n which is wrong anyway.

wolfmanjm commented on src/main.cpp:
```diff
@@ -190,6 +192,14 @@ void init() {
     #ifndef NO_UTILS_MOTORDRIVERCONTROL
     kernel->add_module( new MotorDriverControl(0) );
     #endif
+	#ifndef NO_TOOLS_JOYSTICK
```
whitespace issue

wolfmanjm commented on src/modules/tools/jogger/Jogger.cpp:
```diff
@@ -0,0 +1,292 @@
+    //read the joysticks for each axis and map their values to a target speed
+    for (int c = 0; c < NUM_JOG_AXES; c++) {
+        struct PAD_joystick s;
+        if (PublicData::get_value(joystick_checksum, this->axis_data_source[c], &s)) {
```
Calling this in an interrupt is a very bad idea, this can cause all sorts of bad things to happen. Ideally the joystick and jogger need to be in the same module to avoid this. This MUST be fixed.

skaterzero807 replied:
Is this something that can be moved to the on_main_loop? Ideally I want the jogger and joystick modules to stay separate, so that jogging can be done with any type of device (e.g. encoder) and the joystick can be used in any other module (e.g. feedrate override)

wolfmanjm replied:
you could move it to either on_idle or on_main_loop yes. although continous use will slow things down. it needs to only do this if there are things to process if possible.

skaterzero807 replied:
Would it be safe to have the jogger module explicitly call a "read" function in the joystick module (while in an interrupt) rather than do a Public Data request? I'm not familiar with why the latter is unsafe, so want to make sure before I try the next idea:
Can the slow ticker interrupt just set a flag which tells the pertinent code to be run in the on_main_loop?

wolfmanjm replied:
No it would not be allowed, modules CANNOT call methods in other modules which is why I suggested you merge the two modules.
The general way we do this is to set a flag in the interrupt and then do the read in either on_idle or on_main_loop. as your code issues gcode I suggest you do it in on_main_loop.

skaterzero807 committed:
fix whitespace and gitignore for pull request

skaterzero807 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/Joystick' into feature/Jogger

skaterzero807 committed:
fix newline missing in gitignore

wolfmanjm commented on .gitignore:
```diff
@@ -21,4 +21,4 @@ debug
 /linux_install.log
-/src/default_excludes.mk
\ No newline at end of file
+/src/default_excludes.mk
```
why is this here as there are no changes. generally we do not accept changes to it.

skaterzero807 replied:
I'm not sure, I don't see gitignore listed in the changed files anymore. I think my latest commit (d74823075213b33198c723f2cc3c86c70f6c1cf1) returned the file back to normal

skaterzero807 committed:
add joystick example config

wolfmanjm commented on ConfigSamples/Snippets/joystick-2D.config:
```diff
@@ -0,0 +1,16 @@
+joystick.vertical.auto_zero                  true
+joystick.vertical.startup_time               1000
+joystick.vertical.refresh_rate               100
+joystick.vertical.start_value                0
```
you need to set your editor to add \n to the end of files :)

skaterzero807 committed:
fix joystick example config

skaterzero807 committed:
Merge branch 'feature/Jogger' of https://github.com/skaterzero807/Smoothieware into feature/Jogger

skaterzero807 committed:
add sample configs

skaterzero807 committed:
refactor to avoid public data request in interrupt

* changed code that reads joysticks to occur in the main loop, only if a
flag has been set in the slow ticker
* moved functions that are private into private

wolfmanjm commented:
correct me if I am wrong but I believe the syntax for initializing fields is `int xxx{0`} not `int xxx= 0 ` in the header.

skaterzero807 commented:
I actually have no idea, I found this when looking for the answer:
https://google.github.io/styleguide/cppguide.html#Variable_and_Array_Initialization
If smoothie's preferred syntax is to use curly braces I can make the change and commit that.

wolfmanjm commented:
well to be consistent when it is an initializer in the header we use {}.
We use = when it is a variable in the body of a method.
There are subtle differences.
the {} in the header is a new syntax and used to have to be done in the constructor.

wolfmanjm commented on src/modules/tools/jogger/Jogger.h:
```diff
@@ -0,0 +1,59 @@
+    char axis_letter[NUM_JOG_AXES] = {}; //keeps track of which machine axis is controlled by each joystick axis
+    std::vector<std::string> jog_axes; //stores the list of machine axes controlled by the joystick
+    unsigned int axis_index = 0; //stores which item in the list of machine axis configs is in use
+    bool is_active = false; //keeps track of whether the joystick is being moved
```
to save memory we tend to use this syntax...
```
struct {
 bool is_active:1;
 bool is_jogging:1;
 bool do_reading:1;
};
```
This packs it all into one byte. putting them at the end means they are not padded from the next 4 bytes allocation.
You will need to initialize them in the constructor.

wolfmanjm commented on src/modules/tools/jogger/Jogger.h:
```diff
@@ -0,0 +1,59 @@
+    float dead_zone = 0.002f; //distance from 0 in which small joystick movement is ignored
+    float nonlinearity = 1.0f; //nonlinearity parameter for scaling of joystick values
+
+    float position[NUM_JOG_AXES] = {}; //keeps track of the joystick positions for each axis
```
I do not believe this will initialize the array to 0. there is a special GCC syntax for that or do it in the constructor.

skaterzero807 committed:
fix header initializations to use curly braces

skaterzero807 committed:
enforce initialization of jogging variables to zero

add code to Jogger constructor to initialize position and speed
variables to zero

skaterzero807 committed:
remove uneccessary registration in module loaded

piflixe commented:
Dear @skaterzero807 ! I was absolutely amazed when reading the [docu](http://smoothieware.org/joystick) regarding joystick functionality. I am building a decent CNC machine and want to control it with a smoothieboard and external servo drivers. 
I am currently using the latest cnc stable build:
`edge-4f622ed, Build date: Dec 22 2017 21:38:05, MCU: LPC1769, System Clock: 120MHz`
I assume since in the docu it says joystick is not yet released it is not included in this release? I am not a big coder so I don't feel very confident in compiling smoothieware myself. Could you provide a compiled CNC version of smoothieware so I can test it? I will share my experiences and try helping with this as much as I can! Please feel free to contact me on piela@konstruktiv-berlin.de on this subject.

[EDIT]:
Sorry for the noob stuff. I found that there is a firmware binary included in the github files. So now I am using
```
Build version: edge-75df940, Build date: Jan 18 2017 20:24:49, MCU: LPC1769, System Clock: 120MHz
  CNC Build 3 axis
```
and also edited the config like in the example snippets. However, the joystick  doesn't do any action to the steppers. Is there a specific command to enable the jogger mode?

wolfmanjm commented:
FYI there is a new feature in Smoothie ```void SimpleShell::jog(string parameters, StreamOutput *stream)``` which this could call instead of submitting gcode (or it could submit a $J instead). The advantage is this will make the jogging more responsive than issuing a G1 or G0.

cilynx commented:
Howdy folks -- I'm curious about the status of this one and what I can do to help out.  I've been playing with it on the mill this week and it works great, including proportional speed control and honoring hardware endstops.  Here's a quick video of some very basic 3-axis jogging:

https://www.wolfteck.com/assets/20181117_164143.mp4

(Apologies for the background music -- I was rocking out in the garage while playing with this.)

One wonk I found was that when I hooked the joystick 3rd axis to either P0.2 or P0.3, behavior was pretty nasty -- like the board was still trying to control the pins for other purposes even though I don't have any other references to them in my config.  Moving the axis to P0.26 made things work as expected.

If I'm reading [lpc176x-adc-channels-and-pins](http://smoothieware.org/lpc1769-pin-usage#lpc176x-adc-channels-and-pins) correctly, P0.2 and P0.3 are not actually available for ADC use on the Smoothieboard.  Jim, can you confirm if I'm interpreting that correctly?  If they're indeed not available, we should update the [joystick connection docs](http://smoothieware.org/joystick#connections).

With respect to the new direct jogging feature, I think one of the huge pros of @skaterzero807's existing implementation is that it supports moving multiple axes at one time.  My personal vote would be to move forward with the current implementation as-is and then we can work on multi-axes movement in `SimpleShell::jog` as time permits then port the jogger module here over after that.  

The added responsiveness and more technically correct implementation will be great, but submitting gcode works with tolerable lag and I feel like this is one of those "perfect is the enemy of the good" scenarios.

If y'all agree, @wolfmanjm, @arthurwolf -- what sort of testing would you guys like to see before these modules can be accepted into edge?  I have a few smoothiefied machines in the garage and can likely knock out a good chunk of it.

wolfmanjm commented:
P0.2 and P0.3 are not available for use as explained in the wiki...  http://smoothieware.org/lpc1769-pin-usage
I don't think this is really needed by many people and so am personally in favor of not incorporating this into edge as it takes a huge amount of flash space for maybe one or two users.
I certainly would not add it to edge if it does not use $J.
This is one of those modules that needs to exist in a fork IMO, or at best it could be merged but not built by default.

wolfmanjm commented:
It would also need to be tested on all configurations of smoothie which is very unlikely to happen given the lack of interest in this functionality.

testuser-konstruktiv commented:
Please excuse my beginners' question:
I am trying to build this branch using `git checkout Jogger`. After issuing `make clean all` and copying "firmware.bin" to the SD-Card of my Smoothieboard, the version of the built image is edge-xxx, not jogger-xxx. Double checked this with `strings firmware.bin|egrep "edge|jogger"`. Do you have an idea, what could be wrong with my build setup?

Thank you very much!

arthurwolf commented:
Please verify that your local copy of the code does indeed contain the
different "jogger" files that make it different from edge. This might
simply be an issue with not correctly using git, so you need to be sure you
actually got the right files before exploring anything else. just browse
the files and find something that's specifically from jogger and not from
edge. If you can't, you've got a git issue.

On Tue, Apr 2, 2019 at 4:15 PM testuser-konstruktiv <
notifications@github.com> wrote:

> Please excuse my beginners' question:
> I am trying to build this branch using git checkout Jogger. After issuing make
> clean all and copying "firmware.bin" to the SD-Card of my Smoothieboard,
> the version of the built image is edge-xxx, not jogger-xxx. Double checked
> this with strings firmware.bin|egrep "edge|jogger". Do you have an idea,
> what could be wrong with my build setup?
>
> Thank you very much!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1122#issuecomment-479016159>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFZ7yleQBnhkwe4PVcxXPEpDvvOwnks5vc2XqgaJpZM4L-hIx>
> .
>


-- 
Courage et bonne humeur.

testuser-konstruktiv commented:
Thank you for your fast reply!
It is not an issue as you described, for example the folder
`src/modules/tools/jogger`
exists in my `Jogger` branch and not in `edge`. Checked it once again just now.

wolfmanjm commented:
I suggest you read up on how to use git http://rogerdudler.github.io/git-guide/, and especially github.
```
git checkout -b skaterzero807-feature/Jogger edge
git pull git://github.com/skaterzero807/Smoothieware.git feature/Jogger
```

arthurwolf commented:
I can't actually find the jogger branch, can you give me a link ? I need it
to help you further.

On Tue, Apr 2, 2019 at 4:29 PM testuser-konstruktiv <
notifications@github.com> wrote:

> Thank you for your fast reply!
> It is not an issue as you described, for example the folder
> src/modules/tools/jogger
> exists in my Jogger branch and not in edge. Checked it once again just
> now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1122#issuecomment-479023437>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFZOmdXJ3503F3an_FjlbfRic_yMlks5vc2lggaJpZM4L-hIx>
> .
>


-- 
Courage et bonne humeur.

arthurwolf commented:
Added to the documentation :
http://smoothieware.org/third-party-branches#checking-out-pull-requests

On Tue, Apr 2, 2019 at 4:41 PM Jim Morris <notifications@github.com> wrote:

> I suggest you read up on how to use git
> http://rogerdudler.github.io/git-guide/, and especially github.
>
> git checkout -b skaterzero807-feature/Jogger edge
> git pull git://github.com/skaterzero807/Smoothieware.git feature/Jogger
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1122#issuecomment-479028866>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFaLGjtboXvBAVRVuG7FMIU0dakGbks5vc2wCgaJpZM4L-hIx>
> .
>


-- 
Courage et bonne humeur.

cilynx commented:
@testuser-konstruktiv -- also remember when grepping that things are case sensitive by default.

```
rcw@xps:~/Downloads/Smoothieware$ strings LPC1768/main.bin|egrep "edge|Jogger"
edge-697a164
rcw@xps:~/Downloads/Smoothieware$ git checkout -b skaterzero807-feature/Jogger edge
M	src/version.cpp
Switched to a new branch 'skaterzero807-feature/Jogger'
rcw@xps:~/Downloads/Smoothieware$ git pull git://github.com/skaterzero807/Smoothieware.git feature/Jogger
remote: Enumerating objects: 169, done.
remote: Counting objects: 100% (169/169), done.
remote: Total 407 (delta 169), reused 169 (delta 169), pack-reused 238
Receiving objects: 100% (407/407), 66.52 KiB | 920.00 KiB/s, done.
Resolving deltas: 100% (263/263), completed with 18 local objects.
From git://github.com/skaterzero807/Smoothieware
 * branch              feature/Jogger -> FETCH_HEAD
Auto-merging src/main.cpp
Merge made by the 'recursive' strategy.
 ConfigSamples/Snippets/jogger.config              |  12 ++++++
 ConfigSamples/Snippets/joystick-2D.config         |  16 ++++++++
 src/main.cpp                                      |  10 +++++
 src/modules/tools/jogger/Jogger.cpp               | 318 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/modules/tools/jogger/Jogger.h                 |  61 ++++++++++++++++++++++++++++++
 src/modules/tools/joystick/JoystickAxis.cpp       | 145 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/modules/tools/joystick/JoystickAxis.h         |  48 +++++++++++++++++++++++
 src/modules/tools/joystick/JoystickAxisPool.cpp   |  36 ++++++++++++++++++
 src/modules/tools/joystick/JoystickAxisPool.h     |  16 ++++++++
 src/modules/tools/joystick/JoystickPublicAccess.h |  22 +++++++++++
 10 files changed, 684 insertions(+)
 create mode 100644 ConfigSamples/Snippets/jogger.config
 create mode 100644 ConfigSamples/Snippets/joystick-2D.config
 create mode 100644 src/modules/tools/jogger/Jogger.cpp
 create mode 100644 src/modules/tools/jogger/Jogger.h
 create mode 100644 src/modules/tools/joystick/JoystickAxis.cpp
 create mode 100644 src/modules/tools/joystick/JoystickAxis.h
 create mode 100644 src/modules/tools/joystick/JoystickAxisPool.cpp
 create mode 100644 src/modules/tools/joystick/JoystickAxisPool.h
 create mode 100644 src/modules/tools/joystick/JoystickPublicAccess.h
rcw@xps:~/Downloads/Smoothieware$ make -j8
make[1]: Entering directory '/home/rcw/Downloads/Smoothieware/mbed'
Building src
make[2]: Entering directory '/home/rcw/Downloads/Smoothieware/mbed/src'
Building mbed-lpc1768.mk
Building mbed-lpc11u24.mk
make[3]: Entering directory '/home/rcw/Downloads/Smoothieware/mbed/src'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/home/rcw/Downloads/Smoothieware/mbed/src'
make[3]: Entering directory '/home/rcw/Downloads/Smoothieware/mbed/src'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/home/rcw/Downloads/Smoothieware/mbed/src'
make[2]: Leaving directory '/home/rcw/Downloads/Smoothieware/mbed/src'
make[1]: Leaving directory '/home/rcw/Downloads/Smoothieware/mbed'
Building Smoothie
make[1]: Entering directory '/home/rcw/Downloads/Smoothieware/src'
**NOTE** Excluding modules tools/drillingcycles tools/spindle
Building for device lpc1768
make[2]: Entering directory '/home/rcw/Downloads/Smoothieware/src'
Compiling version.cpp
Compiling main.cpp
Compiling modules/tools/joystick/JoystickAxis.cpp
Compiling modules/tools/joystick/JoystickAxisPool.cpp
Compiling modules/tools/jogger/Jogger.cpp
Linking ../LPC1768/main.elf
Extracting ../LPC1768/main.hex
Extracting ../LPC1768/main.bin
Extracting disassembly to ../LPC1768/main.disasm
   text	   data	    bss	    dec	    hex	filename
 392164	    708	   9876	 402748	  6253c	../LPC1768/main.elf

make[2]: Leaving directory '/home/rcw/Downloads/Smoothieware/src'
make[1]: Leaving directory '/home/rcw/Downloads/Smoothieware/src'
rcw@xps:~/Downloads/Smoothieware$ strings LPC1768/main.bin|egrep "edge|Jogger"
skaterzero807-feature/Jogger-1f0af8f
rcw@xps:~/Downloads/Smoothieware$

testuser-konstruktiv commented:
Thank you for your feedback, that helped a lot!
I was assuming, the feature/Jogger branch in https://github.com/skaterzero807/Smoothieware has to be compiled using the sources existing in that repo. Problem solved, thanks!

wolfmanjm commented:
FYI if anyone is interested in completing this PR $J now supports XYZ axis so it should be modified to use that.

wolfmanjm commented:
This is getting old, and request to use $J functionality has not been done, so closing for now.
